### PR TITLE
Add support for glitchfree reconnection to signal-server

### PR DESCRIPTION
**Pull request by pnts-se-whereby**
PR needed for glitch-free reconnection to signal-server.

Test like this:
* `yarn run test:unit:coverage`
* see signal-server PR for test instructions.

<details>
  <summary>📦 Published PR as canary version: <code>1.5.1--canary.36.7000034888.0</code></summary>
  <br />
  
  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @whereby/jslib-media@1.5.1--canary.36.7000034888.0
  # or 
  yarn add @whereby/jslib-media@1.5.1--canary.36.7000034888.0
  ```
</details>

### CODE REVIEW

File: src/utils/ServerSocket.js [hunk 1]
```diff
@@ -41,6 +47,14 @@ export default class ServerSocket {
         });
     }
 
+    setRtcManager(rtcManager) {
+        if (this.glitchFree) {
+            if (!this._reconnectManager) throw new Error("ReconnectManager is missing");
```
Comment (richtrwhereby) [inline src/utils/ServerSocket.js:52]: I don't see anyway to throw this error if `glitchFree` is true. If we keep this throw then I feel like we should handle that correctly in the `RtcManagerDispatcher.js` file change also. But we should just remove it IMO:

```suggestion
        if (this._reconnectManager) {
```

File: src/utils/ServerSocket.js [hunk 2]
```diff
@@ -99,6 +113,19 @@ export default class ServerSocket {
...
+            if (
+                [
+                    PROTOCOL_RESPONSES.ROOM_JOINED,
+                    PROTOCOL_RESPONSES.CLIENT_LEFT,
+                    PROTOCOL_RESPONSES.NEW_CLIENT,
+                ].includes(eventName)
+            ) {
+                return this._interceptEvent(eventName, handler);
+            }
+        }
```
Comment (richtrwhereby) [inline src/utils/ServerSocket.js:127]:
```suggestion
        const relayableEvents = [
            PROTOCOL_RESPONSES.ROOM_JOINED,
            PROTOCOL_RESPONSES.CLIENT_LEFT,
            PROTOCOL_RESPONSES.NEW_CLIENT
        ];
        
        // Intercept certain events if glitch-free is enabled.
        if (this._reconnectManager && relayableEvents.includes(eventName)) {
            return this._interceptEvent(eventName, handler);
        }
```

File: src/utils/ServerSocket.js [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
 
 const DEFAULT_SOCKET_PATH = "/protocol/socket.io/v4";
 
 /**
  * Wrapper class that extends the Socket.IO client library.
  */
 export default class ServerSocket {
-    constructor(hostName, optionsOverrides) {
+    constructor(hostName, optionsOverrides, glitchFree) {
+        this.glitchFree = glitchFree;
```
Comment (richtrwhereby) [inline src/utils/ServerSocket.js:13]: I don't think we need to store `this.glitchFree`. We either have `this._reconnectManager` or we don't and I think it is enough to just check `if(this._reconnectManager)...` in all cases.

File: src/utils/ServerSocket.js [hunk 4]
```diff
@@ -125,4 +152,16 @@ export default class ServerSocket {
...
         this._socket.off(eventName, handler);
     }
+
+    /**
+     * Intercept event and let ReconnectManager handle them.
+     */
+    _interceptEvent(eventName, handler) {
+        if (!this._reconnectManager) throw new Error("ReconnectMangager is missing");
+
+        this._reconnectManager.on(eventName, handler);
```
Comment (richtrwhereby) [inline src/utils/ServerSocket.js:162]: If we are going to throw an error we should have some error handling in the caller IMO. Instead, I would simply remove this and replace with a true check instead:

```suggestion
        if(this._reconnectManager) {
            this._reconnectManager.on(eventName, handler);
        }

```

File: tests/utils/ServerSocket.spec.js [hunk 5]
```diff
@@ -39,4 +43,74 @@ describe("ServerSocket", () => {
...
+
+            expect(serverSocket._reconnectManager?.rtcManager).toBe(mockManager);
+        });
+
+        it("should throw with glitchFree on and missing ReconnectManger", () => {
+            const serverSocket = new ServerSocket("https://localhost", null, true);
+
+            delete serverSocket._reconnectManager;
+
+            expect(() => serverSocket.setRtcManager()).toThrow();
```
Comment (richtrwhereby) [inline tests/utils/ServerSocket.spec.js:73]: If we decide to no longer throw unhandled errors here (see previous comments) then we should update all `toThrow()` and `not.toThrow()` tests here.

**Review by richtrwhereby — CHANGES_REQUESTED** [review]

File: src/model/protocol.js [hunk 6]
```diff
@@ -133,3 +133,7 @@ export const KNOCK_MESSAGES = {
         REJECTED: "rejected",
     },
 };
+
+export const PROTOCOL_EVENTS = {
```
[the remark this answers is not in the document]
Comment (pnts-se-whereby) [reply src/model/protocol.js:137]: To me, it didn't really fit into the other param types. If you think it does let me know and I'll change, else I'll add a comment.

File: src/utils/ServerSocket.js (hunk 3 above)
Comment (pnts-se-whereby) [reply src/utils/ServerSocket.js:13]: Agree.

File: tests/utils/ServerSocket.spec.js (hunk 5 above)
Comment (pnts-se-whereby) [reply tests/utils/ServerSocket.spec.js]: Done

File: src/utils/ReconnectManager.js [hunk 7]
```diff
@@ -0,0 +1,264 @@
+import EventEmitter from "events";
+import { getUpdatedStats } from "../webrtc/stats/StatsMonitor/index";
+import { PROTOCOL_EVENTS, PROTOCOL_RESPONSES } from "../model/protocol";
+
+export const PEER_CONNECTION_RENEW_THRESHOLD = 2500; // reconnects completed withing this threshold can be "glitch-free"
```
[the remark this answers is not in the document]
Comment (pnts-se-whereby) [reply src/utils/ReconnectManager.js]: I changed the reconnect_threshold when re-joining according to our discussion.

File: src/model/protocol.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (richtrwhereby) [reply src/model/protocol.js:138]: Let's just add a comment explaining what this constant is and why it is separated (see other constants in this file for examples).

File: src/utils/ReconnectManager.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (richtrwhereby) [reply src/utils/ReconnectManager.js]: Could you explain what `reconnect_threshold` is?

Also, there is no change to the comment here to explain the interrelationship of this parameter with other parameters e.g. `disconnectTimeout`?

[the remark this answers is not in the document]
Comment (pnts-se-whereby) [reply src/utils/ReconnectManager.js]: Reconnect threshold is used for deciding if we should try glitch-free reconnection. It is calculated like this: `disconnectTimeout * 0.8`. I updated the comment in the code where the threshold is being used, i.e. when intercepting the `room_joined` event.
Fixed in https://github.com/whereby/jslib-media/pull/36/commits/6576753d3facf2a441c9cb22fd3da15a4928e615

[the remark this answers is not in the document]
Comment (richtrwhereby) [reply src/utils/ReconnectManager.js]: Looks great! 🎉

**Review by richtrwhereby — APPROVED** [review]

File: package.json [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@whereby/jslib-media",
   "description": "Media library for Whereby",
-  "version": "1.4.4",
+  "version": "1.5.0",
```
Comment (richtrwhereby) [inline package.json:4]: Note: Before merge please double check this value is the correct bump to `main` version (i.e. no other bumps were made on the `main` branch).

Comment (pnts-se-whereby) [reply package.json:4]: done

**Pull request merged**